asm/instruction_test.go

```diff
@@ -280,3 +288,46 @@ func TestInstructionIterator(t *testing.T) {
 		}
 	}
 }
+
+func TestMetadataCopyOnWrite(t *testing.T) {
+	insn := Instructions{
+		JEq.Imm(R1, 123, "my_func"),
+		Mov.Reg(R2, R1),
+		Mul.Reg(R1, R2),
+	}
+
+	insn2 := make([]Instruction, len(insn))
+	copy(insn2, insn)
+
+	// Setting the reference should have copied the metadata object and not effected the exiting pointer
+	insn2[0].SetReference("my_func2")
+
+	if insn[0].Reference() != "my_func" {
+		t.Fatal("metadata modification to copied instruction effected the old instruction")
+	}
+
+	if insn2[0].Reference() != "my_func2" {
+		t.Fatal("SetReference didn't update new instruction")
+	}
+
+	if insn[0].metadata == insn2[0].metadata {
+		t.Fatal("changed instruction should not be equal")
+	}
+
+	// Set the reference, then clear it. Causing us to have a nil in one instruction and an empty metadata in the
+	// other.
+	insn[1].SetReference("SomeValue")
+	insn[1].SetReference("")
+
+	// Metadata is value compared, not pointer compared, so so these should still be equal
+	if !insn[1].equal(insn2[1]) {
```

Comment:
`equal()` no longer compares metadata.

Reply:
Right, bit of a theme, have to update the unit tests after the changed made during the review